### ISSUE TITLE
OpenOCD broken by newest libusb (mingw-w64-x86_64-openocd-0.11.0-1 + mingw-w64-x86_64-libusb-1.0.25-2)

### ISSUE THREAD

**Issue by radioflash** [msys2/MINGW-packages#10843]
A recent update broke OpenOCD for me.
I narrowed it down to a segfault in libusb-- replacing libusb-1.0.dll (from the "mingw-w64-x86_64-libusb-1.0.25-2"-package) with an older version of the file circumvents the problem for me, no other change is necessary (reinstalling libusb with `pacman -S mingw-w64-x86_64-libusb` also breaks things again).

This is an insiduous problem because it results in a segfault ONLY when actually connecting to the debugger.
Just running openocd.exe without really doing anything still works fine:

> C:\msys64\mingw64\bin>openocd.exe --version
> Open On-Chip Debugger 0.11.0
> Licensed under GNU GPL v2
> For bug reports, read
>         http://openocd.org/doc/doxygen/bugs.html

This also still works when the debugger is not connected:

> $ openocd -f openocd-chip-1.cfg -c "targets stm32f7x.cpu"
> Open On-Chip Debugger 0.11.0
> Licensed under GNU GPL v2
> For bug reports, read
>         http://openocd.org/doc/doxygen/bugs.html
> Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
> Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
> DEPRECATED! use 'adapter speed' not 'adapter_khz'
> adapter speed: 8000 kHz
> 
> Info : Listening on port 6666 for tcl connections
> Info : Listening on port 4444 for telnet connections
> Info : clock speed 8000 kHz
> Error: open failed

But when the debugger is connected, then it segfaults:

> (gdb) r
> Starting program: C:\msys64\mingw64\bin\openocd.exe -f openocd-chip-1.cfg -c "targets stm32f7x.cpu"
> [New Thread 8592.0x3004]
> [New Thread 8592.0x4658]
> [New Thread 8592.0x4d80]
> Open On-Chip Debugger 0.11.0
> Licensed under GNU GPL v2
> For bug reports, read
>         http://openocd.org/doc/doxygen/bugs.html
> Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
> Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
> DEPRECATED! use 'adapter speed' not 'adapter_khz'
> adapter speed: 8000 kHz
> 
> Info : Listening on port 6666 for tcl connections
> Info : Listening on port 4444 for telnet connections
> Info : clock speed 8000 kHz
> [New Thread 8592.0xfc0]
> 
> Thread 1 received signal SIGSEGV, Segmentation fault.
> 0x00007ffbfd08761d in ?? () from C:\msys64\mingw64\bin\libusb-1.0.dll

Workaround for now:
Replacing `mingw64/bin/libusb-1.0.dll` with an older version fixes the problem.

Version info for libusb:
"Good" libusb DLL (mingw-w64-x86_64-libusb binary that works, but is significantly older):
major, minor, micro, nano, patch, rc, describe: 1, 0, 20, 11004, , http://libusb.info
"Bad" libusb (currently shipped mingw-w64-x86_64-libusb):
1, 0, 25, 11692, , http://libusb.info

I suspect that this was caused by the latest update to libusb, but I'm unsure WHY this happens or what exactly goes wrong.

Possible next step: Build libus with debug symbols to see where it segfaults?

Any advice on helpful next steps would be very welcome-- I'd love to get this fixed "properly" and just downgrading libusb is probably not a pretty/proper solution for everyone.

**Comment by lazka** [comment]
Does this mean 1.0.24 also is broken? 1.0.20 is 7 years old

**Comment by radioflash** [comment]
> Does this mean 1.0.24 also is broken?

I just tested, and mingw-w64-x86_64-libusb-1.0.24-5 *definitely* works:
`makepkg -sCLf` on  commit 7c67d9c46 produces a working mingw-w64-x86_64-libusb-1.0.24-5-any.pkg.tar.zst,
while installing mingw-w64-x86_64-libusb-1.0.25-2-any.pkg.tar.zst built from the current master causes the segfault again.

**Comment by radioflash** [comment]
Ok found out more:
TL;DR: It's Openocd that's broken, they fixed it in June 2021, but the 0.11.0 release is older (the fix is currently in OpenOCDs git repo only).

How to resolve this?
1) Should I pester the OpenOCD guys for a new release (not sure if feasible)?
2) Should I try and integrate a patch for this issue only that works for openod 0.11.0 and commit that here? (it would probably need to be touched later, on the next openocd releases).
3) Could the OpenOCD package be changed into using a git version of OpenOCD?

Details:
OpenOCD crashes in `libusb_get_next_timeout` because its internal default context (usbi_default_context) is not set.
But this is OpenOCDs fault, since it calls `libusb_init` with a non-default context, but then tries to operate libusb with the default context (NULL) later.
This used to work before libusb 1.0.25, even though it was API-misuse.

See discussion on the libusb project (OpenOCD issue is also linked):
https://github.com/libusb/libusb/issues/928

**Comment by Biswa96** [comment]
I think option 2 looks better. Adding the patch from upstream would have minimal effect in other aspects.

> it would probably need to be touched later, on the next openocd release

A comment can be added above the patch command for future notice.

**Comment by Biswa96** [comment]
This issue is fixed in the above linked pull request. The fixed version (0.11.0-2) can be seen here https://packages.msys2.org/base/mingw-w64-openocd
